src/main/java/com/yahoo/bullet/bql/query/QueryValidator.java

```diff
@@ -0,0 +1,197 @@
+        // If the query type is select distinct, the select nodes are the actual aggregate nodes.
+        ProcessedQuery.QueryType queryType = processedQuery.getQueryType();
+
+        if (queryType == ProcessedQuery.QueryType.SELECT_DISTINCT) {
+            expressionValidator.process(processedQuery.getSelectNodes().stream().map(SelectItemNode::getExpression).collect(Collectors.toList()), processedQuery.getPreAggregationMapping());
+        } else {
+            expressionValidator.process(processedQuery.getAggregateNodes(), processedQuery.getPreAggregationMapping());
+            expressionValidator.process(processedQuery.getGroupByNodes(), processedQuery.getPreAggregationMapping());
+        }
+
+        // Replace schema with aggregate fields
+        List<Schema.Field> aggregateFields = null;
+        switch (queryType) {
+            case SELECT_DISTINCT:
+                aggregateFields = processedQuery.getSelectNodes().stream()
+                                                                 .map(SelectItemNode::getExpression)
+                                                                 .map(toSchemaField(processedQuery))
+                                                                 .collect(Collectors.toList());
+                break;
+            case GROUP:
+                aggregateFields = Stream.concat(processedQuery.getAggregateNodes().stream(),
+                                                processedQuery.getGroupByNodes().stream())
+                                        .map(toSchemaField(processedQuery))
+                                        .collect(Collectors.toList());
+                break;
+            case COUNT_DISTINCT:
+                aggregateFields = toSchemaFields(processedQuery.getCountDistinct(), processedQuery);
+                break;
+            case DISTRIBUTION:
+                schema.replaceSchema(DISTRIBUTION_SCHEMAS.get(query.getAggregation().getAttributes().get(Distribution.TYPE)));
+                break;
+            case TOP_K:
+                aggregateFields = toSchemaFields(processedQuery.getTopK().getExpressions(), processedQuery);
+                aggregateFields.add(new Schema.PlainField(getTopKName(query), Type.LONG));
+                break;
+            case SPECIAL_K:
+                aggregateFields = toSchemaFields(processedQuery.getGroupByNodes(), processedQuery);
+                aggregateFields.add(new Schema.PlainField(getTopKName(query), Type.LONG));
+                break;
+        }
+        if (aggregateFields != null) {
+            duplicates(aggregateFields).ifPresent(duplicates -> {
+                    processedQuery.getErrors().add(new BulletError("The following field names/aliases are shared: " + duplicates, null));
+                });
+            schema.replaceSchema(new Schema(aggregateFields));
+        }
+
+        // If there's a HAVING clause, check that it can be evaluated as a boolean. The HAVING clause in special k queries doesn't count.
```

Comment:
HAVING clause is covered by the aggregation. There's no actual HAVING postaggregation in the query for special k.It might be ok to check it anyways or it might throw an error. not sure.